src/full/Agda/TypeChecking/Records.hs

```diff
@@ -697,6 +698,73 @@ isSingletonType' regardIrrelevance t = do
           emap (abstract tel) <$> isSingletonRecord' regardIrrelevance r ps
         _ -> return $ Right Nothing
 
+-- | Checks whether the given term (of the given type) is beta-eta-equivalent
+--   to a variable. Returns just the de Bruijn-index of the variable if it is,
+--   or nothing otherwise.
+isEtaVar :: Term -> Type -> TCM (Maybe Int)
```

Comment:
We are occasionally (but far from always) using `Nat` instead of `Int` for deBruijn indices. If we want to introduce a new alias (or `newtype`?), it should be done in a separate refactoring from this PR.

Reply:
Yes, we should get rid of `Int` in favour of a type synonym or newtype.
Maybe someone volunteers for this refactoring, otherwise we can also do-as-we-go.

Reply:
@UlfNorell : Please do a review, not just a meta-review! ;-)

Reply:
> Yes, we should get rid of `Int` in favour of a type synonym or newtype.

Since we only support GHC >= 7.10.(3), what about of using  the `Natural` data type from [Numeric.Natural](https://downloads.haskell.org/~ghc/7.10.1/docs/html/libraries/Numeric-Natural.html)?

Reply:
Using a machine word rather than a bigint for deBruijn indices is actually gives a noticeable performance improvement. I don't remember the actual numbers, but I remember it was non-trivial.

Maybe we should have this discussion in a separate issue instead of in Jesper's completely unrelated PR..?